casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala

```diff
@@ -173,16 +172,34 @@ object BlockApproverProtocol {
             vaults,
             Long.MaxValue
           )
-          .toSet
         blockDeploys = block.body.deploys
         _ <- (blockDeploys.size == genesisBlessedContracts.size)
               .either(())
               .or("Mismatch between number of candidate deploys and expected number of deploys.")
+
+        // check if expected blessed contracts and ones reported by candidate are equal
+        // order should be the same and terms should be the same
+        wrongDeploys = (blockDeploys zip genesisBlessedContracts)
+          .mapFilter {
+            case (candidateDeploy, expectedContract) => {
+              val term    = candidateDeploy.deploy.data.term
+              val isWrong = term != expectedContract.data.term
+              if (isWrong) term.lines.findFirst.get.some else none[String]
+            }
+          }
```

Comment:
To better understand what is executed, try this experiment. I'm interested what would you change in your code.

```scala
val items = 1 to 10

object test1 {
  val results = items
    .filter(_ < 5)
    .take(2)
    .map { x => println(s"Test 1 - number $x"); x }

  println(s"Test 1: $results\n")
}

object test2 {
  val results = items
    .filter(_ < 5)
    .map { x => println(s"Test 2 - number $x"); x
    }
    .take(2)

  println(s"Test 2: $results\n")
}

object test3 {
  import cats.syntax.all._

  val results = items.toList
    .mapFilter { x =>
      println(s"Test 3 - number $x")
      if (x < 5) x.some else none
    }
    .take(2)

  println(s"Test 3: $results\n")
}

test1; test2; test3
```

Reply:
I'm not sure what do you mean. I have to calculate `isWrong` for all pairs and pass down inconsistencies.

Reply:
It can be many deploys so we don't need to check all of them. We can fail on first but also we can display three for example, as I described in my initial comment.
Also, one line can hold large string so that's why it make sense to extract only prefix just to recognize which deploy does not match.

What I mean by my sample code is to show how using List (Seq) and Stream is different and it is advisable to check what is exactly running to eliminate unnecessary executions.

Reply:
`we can display three for example` - you mean three inconsistent? how do we get them if we fail fast on first?
So I assume this will stop traversing `(blockDeploys zip genesisBlessedContracts)` List on first failure?
```scala 
    wrongDeploys = (blockDeploys zip genesisBlessedContracts)
      .filter {
        case (candidateDeploy, expectedContract) => {
          candidateDeploy.deploy.data.term != expectedContract.data.term
        }
      }
      .take(1)
```

Reply:
Yes, we can choose how many items will be returned, one or three, after which iteration will stop.

Now, if we add one small change and use `Iterator` instead of the collection directly, behavior will be more like lazy without allocating intermediate collections. You will notice this is very similar to fs2 Pure streams.
```scala
object test2_iterator {
  val results = items.iterator
    .filter(_ < 5)
    .map { x => println(s"Test 2 iter - number $x"); x
    }
    .take(2)

  println(s"Test 2 iter: ${results.toList}\n")
}
```
Maybe you will remember @marcin-rzeznicki's PR #2812 with optimizations using iterators.